File: solutions/haskell/sum-of-multiples/1/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

import Data.Set (Set)
import qualified Data.Set as Set

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples factors limit = sum . Set.fromList $ factors >>= \x -> multiples x limit


multiples :: Integer -> Integer -> [Integer]
multiples n limit 
 | n > 0 = [n,n+n..limit-1]
 | otherwise = [0]

File: solutions/scala/sum-of-multiples/1/src/main/scala/SumOfMultiples.scala
object SumOfMultiples {
  def sum(factors: Set[Int], limit: Int): Int = {
    factors
      .flatMap(n => multiples(n, limit))
      .sum
  }

  def multiples(n: Int, limit: Int): List[Int] = (n until limit by n).toList
}

